BMEdit/Editor/Source/Widgets/SceneRenderWidget.cpp

```diff
@@ -993,218 +1016,302 @@ namespace widgets
+			{
+			case gamelib::mat::MATBlendMode::BM_TRANS:
+				gapi->glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+				break;
+			case gamelib::mat::MATBlendMode::BM_TRANS_ON_OPAQUE:
+				gapi->glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+				break;
+			case gamelib::mat::MATBlendMode::BM_TRANSADD_ON_OPAQUE:
+				gapi->glBlendFunc(GL_SRC_ALPHA, GL_ONE);
+				break;
+			case gamelib::mat::MATBlendMode::BM_ADD_BEFORE_TRANS:
+				gapi->glBlendFunc(GL_ONE, GL_ONE);
+				break;
+			case gamelib::mat::MATBlendMode::BM_ADD_ON_OPAQUE:
+				gapi->glBlendFunc(GL_ONE, GL_ONE);
+				break;
+			case gamelib::mat::MATBlendMode::BM_ADD:
+				gapi->glBlendFunc(GL_ONE, GL_ONE);
+				break;
+			default:
+				// Do nothing
+				break;
+			}
 
-					// 4. Render mesh
-					if (m_renderMode & RenderMode::RM_TEXTURE)
-					{
-						// normal draw
-						mesh.render(glFunctions);
-					}
+			// Enable or disable alpha testing
+			if (renderState.isAlphaTestEnabled()) {
+				gapi->glEnable(GL_ALPHA_TEST);
+			} else {
+				gapi->glDisable(GL_ALPHA_TEST);
+			}
 
-					if (m_renderMode & RenderMode::RM_WIREFRAME)
-					{
-						glFunctions->glPolygonMode(GL_FRONT_AND_BACK, GL_LINE);
-						mesh.render(glFunctions);
-						// reset back
-						glFunctions->glPolygonMode(GL_FRONT_AND_BACK, GL_FILL);
-					}
+			// Enable or disable fog
+			if (renderState.isFogEnabled()) {
+				gapi->glEnable(GL_FOG);
+			} else {
+				gapi->glDisable(GL_FOG);
+			}
 
-					// 5. Unbind texture and shader (expected to switch between materials, but not now)
-					glFunctions->glBindTexture(GL_TEXTURE_2D, 0);
-					texturedShader.unbind(glFunctions);
-				}
+			// Enable or disable depth offset (Z bias)
+			if (renderState.hasZBias()) {
+				gapi->glEnable(GL_POLYGON_OFFSET_FILL);
+				gapi->glPolygonOffset(1.0f, renderState.getZOffset());
+			} else {
+				gapi->glDisable(GL_POLYGON_OFFSET_FILL);
+			}
+
+			// Set cull mode based on your enum values
+			switch (renderState.getCullMode())
+			{
+			case gamelib::mat::MATCullMode::CM_OneSided:
+				gapi->glCullFace(GL_BACK);
+				break;
+			case gamelib::mat::MATCullMode::CM_DontCare:
+			case gamelib::mat::MATCullMode::CM_TwoSided:
+				// please complete
+				gapi->glDisable(GL_CULL_FACE);
+				break;
 			}
+		};
+
+		static constexpr std::array<std::string_view, render::TextureSlotId::kMaxTextureSlot> g_aTextureKindToLocation {
+		    "i_uMaterial.mapDiffuse",
+		    "i_uMaterial.mapSpecularMask",
+		    "i_uMaterial.mapEnvironment",
+		    "i_uMaterial.mapReflectionMask",
+		    "i_uMaterial.mapReflectionFallOff",
+		    "i_uMaterial.mapIllumination",
+		    "i_uMaterial.mapTranslucency"
+		};
+
+		for (const auto& entry : entries)
+		{
+			if (!filter(entry))
+				continue; // skipped by filter
+
+			// Switch render state
+			applyRenderState(glFunctions, entry.material.renderState);
+
+			// Activate material & setup parameters
+			render::Shader* shader = entry.material.pShader;
+			shader->bind(glFunctions);
```

Comment:
We can try to optimize this too: we able to group objects by shaders and decrease bind/unbind operations. Literally it's not required now, but could be useful in future 🤔